File: src/cc/tree/MaximumWidthOfBinaryTree.cc
#include <queue>

using namespace std;

/**
 * Definition for a binary tree node.
 */
class TreeNode {
public:
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

/**
 * LeetCode 662. Maximum Width of Binary Tree
 * https://leetcode.com/problems/maximum-width-of-binary-tree/
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        int maximumWidth = 0;
        queue<TreeNode*> aQueue;
        // BigInt problem...
        queue<long long> aQueueIndex;
        aQueue.push(root);
        aQueueIndex.push(1);
        while (!aQueue.empty()) {
            int size = aQueue.size();
            long long widthStart = 0;
            long long widthEnd = 0;
            for (int i = 0; i < size; i++) {
                TreeNode *currentNode = aQueue.front();
                long long currentIndex = aQueueIndex.front();
                aQueue.pop();
                aQueueIndex.pop();
                if (i == 0) {
                    widthStart = currentIndex;
                }
                if (i == size - 1) {
                    widthEnd = currentIndex;
                }
                if (currentNode -> left != nullptr) {
                    aQueue.push(currentNode -> left);
                    long long nextIndex = 2 * currentIndex;
                    aQueueIndex.push(nextIndex);
                }
                if (currentNode -> right != nullptr) {
                    aQueue.push(currentNode -> right);
                    long long nextIndex = 2 * currentIndex + 1;
                    aQueueIndex.push(nextIndex);
                }
            }
            int currentWidth = widthEnd - widthStart + 1;
            maximumWidth = currentWidth > maximumWidth ? currentWidth : maximumWidth;
        }
        return maximumWidth;
    }
};
int main(int argc, char const *argv[]) {
    // ...
    return 0;
}
/* EOF */

File: src/io/ziheng/tree/leetcode/MaximumWidthOfBinaryTree.java
package io.ziheng.tree.leetcode;

import java.util.Queue;
import java.util.LinkedList;

/**
 * Definition for a binary tree node.
 */
class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode() {
        // ...
    }
    public TreeNode(int val) {
        this.val = val;
    }
    public TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

/**
 * LeetCode 662. Maximum Width of Binary Tree
 * https://leetcode.com/problems/maximum-width-of-binary-tree/
 */
public class MaximumWidthOfBinaryTree {
    public static void main(String[] args) {
        // ...
    }
    public int widthOfBinaryTree(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int maximumWidth = 0;
        Queue<TreeNode> aQueue = new LinkedList<>();
        Queue<Integer> aQueueIndex = new LinkedList<>();
        aQueue.offer(root);
        aQueueIndex.offer(1);
        while (!aQueue.isEmpty()) {
            int size = aQueue.size();
            int widthStart = 0;
            int widthEnd = 0;
            for (int i = 0; i < size; i++) {
                TreeNode currentNode = aQueue.poll();
                int currentIndex = aQueueIndex.poll();
                if (i == 0) {
                    widthStart = currentIndex;
                }
                if (i == size - 1) {
                    widthEnd = currentIndex;
                }
                if (currentNode.left != null) {
                    aQueue.offer(currentNode.left);
                    aQueueIndex.offer(2 * currentIndex);
                }
                if (currentNode.right != null) {
                    aQueue.offer(currentNode.right);
                    aQueueIndex.offer(2 * currentIndex + 1);
                }
            }
            maximumWidth = Math.max(maximumWidth, widthEnd - widthStart + 1);
        }
        return maximumWidth;
    }
}

/* EOF */